lib/spack/spack/variant.py

```diff
@@ -66,11 +66,26 @@ def __init__(
         self.default = default
         self.description = str(description)
 
+        self.values = None
+        if values is any:
+            # 'any' is a special case to make it easy to say any value is ok
+            self.single_value_validator = lambda x: True
+
+        elif isinstance(values, type):
+            # supplying a type means any value *of that type*
+            def isa_type(v):
+                try:
+                    values(v)
+                    return True
+                except ValueError:
+                    return False
+            self.single_value_validator = values
```

Comment:
I think there may be a bug on this not tested line. Shouldn't this line read:
```python
self.single_value_validator = isa_type
```
? Also, the function `isa_type` assumes that we have a `values.__init__`  method that takes a value of the same type. What about using:
```python
 elif isinstance(values, type):
    self.single_value_validator = lambda x: isinstance(x, values)
```
instead of the function above?